README.rst
Distributed tests planner plugin for pytest testing framework
=============================================================

.. image:: https://img.shields.io/pypi/v/pytest-cloud.svg
   :target: https://pypi.python.org/pypi/pytest-cloud
.. image:: https://img.shields.io/pypi/pyversions/pytest-cloud.svg
  :target: https://pypi.python.org/pypi/pytest-cloud
.. image:: https://img.shields.io/coveralls/pytest-dev/pytest-cloud/master.svg
   :target: https://coveralls.io/r/pytest-dev/pytest-cloud
.. image:: https://travis-ci.org/pytest-dev/pytest-cloud.svg?branch=master
    :target: https://travis-ci.org/pytest-dev/pytest-cloud
.. image:: https://readthedocs.org/projects/pytest-cloud/badge/?version=latest
    :target: https://readthedocs.org/projects/pytest-cloud/?badge=latest
    :alt: Documentation Status


Install pytest-cloud
--------------------

::

    pip install pytest-cloud

    # install GNU parallel utility
    # https://www.gnu.org/software/parallel/
    # for MacOS, you can use:
    brew install parallel


.. _pytest: http://pytest.org
.. _pytest-xdist: https://pypi.python.org/pypi/pytest-xdist


Features
--------

The plugin provides an easy way of running tests among several test nodes (workers).
Uses the great pytest-xdist_ plugin for actual distributed run.
When used, it will automatically detect capabilites of given node(s) and run only the number of test processes it is
able to handle. If will also filter out offline nodes or nodes which were failed to respond to the
capabilities acquisition call.

Supports automatic codebase propagation to the test nodes, so you don't have to install python dependencies
for your project on remote test nodes globally - just make sure that your virtualenv folder is `inside`
your project folder - that's a requirement.
It will also detect a root folder of the test environment (project root), and will `rsync` it to all test nodes.

NOTE:
> `pytest-cloud` uses `virtualenv` instead of built-in `venv` package.


ATM only ssh transport is supported. So ensure that you have at least public key auth enabled to your test nodes
from the master node (where py.test is executed).


Command-line options
--------------------

* `--cloud-node`
    Node hostname (or user@hostname) to run tests on. Multiple allowed.

* `--cloud-nodes`
    Space-separated list of node hostname (or user@hostname) to run tests on. Multiple allowed.

* `--cloud-python`
    Optional python executable name to be used on the remote test nodes.
    Default is the executable name used for the test run on the master.

* `--cloud-chdir`
    Optional relative path to be used on the remote test nodes as target folder for syncing file and run tests.
    Default is `pytest_<username>_<current_folder_name>`.

* `--cloud-virtualenv-path`
    Optional relative path to the virtualenv to be used on the remote test nodes. By default it will try to detect
    whether current test process is using virtualenv and if it's located inside of the current directory. If that's
    the case, it will use it for rsync on the remote node(s).

* `--cloud-mem-per-process`
    Optional amount of memory roughly needed for test process, in megabytes.
    Will be used to calculate amount of test processes per node, getting the free memory, dividing it for the memory
    per process needed, and getting the minimum of that value and the number of CPU cores of the test node.

* `--cloud-max-processes`
    Optional maximum number of processes per test node. Overrides from above the calculated number
    of processes using memory and number of CPU cores.

* `--cloud-rsync-bandwidth-limit`
    Optional bandwidth limit per `rsync` process, in kilobytes per second. 5000 by default.

* `--cloud-rsync-max-processes`
    Optional process count limit for `rsync` processes. By default there's no limit so rsyncing will be in parallel
    for all test nodes.

* `--cloud-rsync-cipher`
    Optional ssh cipher selection for `rsync` processes. aes128-gcm@openssh.com by default.
    Default cipher is chosen to have the least possible network overhead. Network overhead is system, compilation
    and CPU architecture dependent, however chosen cipher is showing good results in majority of use cases.

Ini file options
----------------

* `cloud_develop_eggs`
    Optional list of python package paths to install in development mode on remote side. Required to be inside of the
    project root directory.


Example
-------

.. code-block:: sh

    py.test tests/ --cloud-node=10.0.120.{1..40} --cloud-mem-per-process=1000 --rsyncdir=.

Or if you pass list of nodes as space-separated list:

.. code-block:: sh

    py.test tests/ --cloud-nodes='10.0.120.1 10.0.120.2' --cloud-mem-per-process=1000 --rsyncdir=.


Contact
-------

If you have questions, bug reports, suggestions, etc. please create an issue on
the `GitHub project page <http://github.com/pytest-dev/pytest-cloud>`_.


License
-------

This software is licensed under the `MIT license <http://en.wikipedia.org/wiki/MIT_License>`_

See `License file <https://github.com/pytest-dev/pytest-cloud/blob/master/LICENSE.txt>`_


© 2015 Anatoly Bubenkov and others.


CHANGES.rst
Changelog
=========

5.0.3
-----

- Fixes issue #21 (dasm)
- Fully dropped python2 support (dasm)
- Added explicit requirement for virtualenv (dasm)

5.0.2
-----

- Remove python2 support (bubenkoff)


5.0.1
-----

- Correctly handle default venv location (bubenkoff)


5.0.0
-----

- Compatible with pytest-xdist >= 1.26.0 *including* 2.0.0+ (Zac-HD)

4.0.0
-----

- Add config option to select cipher for ssh connection (dasm)
- Fix deprecation warnings for pytest (dasm)
- Support pytest-xdist >= 1.26.0 (dasm)
- Support pytest >= 3.6.1 (dasm)
- Remove support for Python 3.0, 3.1, 3.2, 3.3 (consistent with pytest-xdist) (dasm)

3.0.1
-----

- support python 3.7 (bubenkoff)

3.0.0
-----

- support pytest-xdist >=1.22.1 (bubenkoff)

2.0.0
-----

- pytest fixed version number is removed from the requirements (olegpidsadnyi)
- removed pytest-cache dependency (olegpidsadnyi)

1.3.8
-----

- Add verbosity to rsync progress (bubenkoff)

1.3.4
-----

- Add develop eggs setting to install packages on remote side in development mode (bubenkoff)

1.2.16
------

- Correctly handle python path on remote side (bubenkoff)

1.2.12
------

- Add rsync progress to the output, change default bandwidth limit (bubenkoff)

1.2.11
------

- Fast native rsync instead of python based one (bubenkoff)

1.1.0
-----

- Avoid unnecessary multiple ssh connections to the same node (bubenkoff)

1.0.25
------

- Correct virtualenv execution (bubenkoff)

1.0.18
------

- Ensure plugin command line hook is executed first (bubenkoff)
- Correct free memory calculation (bubenkoff)

1.0.15
------

- Add pyc files cleanup (bubenkoff)

1.0.13
------

- Automatic discovery of the virtualenv (bubenkoff)
- Fixes to rsyncing (bubenkoff)
- Safer node id generation (bubenkoff)
- Guarantee uniqueness of provided nodes (bubenkoff)

1.0.10
------

- Delete orphan files when rsyncing (bubenkoff)


1.0.7
-----

- Add possibility to pass node list as space separated (bubenkoff)


1.0.6
-----

- Initial public release


tests/conftest.py
"""Configuration for pytest runner."""

# pylint: disable=C0103
pytest_plugins = "pytester"


tests/test_plugin.py
"""Tests for pytest-bdd-splinter subplugin."""
import sys

import mock
import pytest

import xdist.dsession
import execnet
import pytest_cloud.plugin


PYTHON = "python{0}.{1}".format(*sys.version_info)


@pytest.fixture
def mocked_group(request):
    """Mock execnet.Group."""
    group = execnet.Group

    def fin():
        execnet.Group = group

    mocked_group = mock.Mock()
    request.addfinalizer(fin)
    mocked_group.mkgateway.return_value = mock.Mock()
    execnet.Group = mocked_group
    return mocked_group


@pytest.fixture
def mocked_dsession(request):
    """Mock xdist.dsession.DSession."""
    dsession = xdist.dsession.DSession

    def fin():
        xdist.dsession.DSession = dsession

    mocked_dsession = mock.Mock()
    request.addfinalizer(fin)

    def iterate():
        yield
        yield

    def dic():
        return {}

    def rep():
        # pylint: disable=R0903,R0205
        class Report(object):
            skipped = []
            call = []
            failed = []
            result = []
            passed = True
            outcome = "OK"
            nodeid = 123

        return Report()

    mocked_dsession.return_value.pytest_addhooks = iterate
    mocked_dsession.return_value.pytest_namespace = dic
    mocked_dsession.return_value.pytest_addoption = iterate
    mocked_dsession.return_value.pytest_configure = iterate
    mocked_dsession.return_value.pytest_collectstart = iterate
    mocked_dsession.return_value.pytest_make_collect_report = rep
    mocked_dsession.return_value.pytest_collectreport = iterate
    mocked_dsession.return_value.pytest_collection_modifyitems = iterate

    xdist.dsession.DSession = mocked_dsession
    return mocked_dsession


@pytest.fixture
def mocked_rsync(request):
    """Mock pytest_cloud.plugin.RSync."""
    rsync = pytest_cloud.plugin.RSync

    def fin():
        pytest_cloud.plugin.RSync = rsync

    request.addfinalizer(fin)

    mocked_rsync = mock.Mock()
    pytest_cloud.plugin.RSync = mocked_rsync

    return mocked_rsync


@pytest.mark.parametrize(
    [
        "host1",
        "user1",
        "cpu_count1",
        "memory1",
        "host2",
        "user2",
        "cpu_count2",
        "memory2",
        "mem_per_process",
        "max_processes",
        "result",
    ],
    [
        (
            "1.example.com",
            "",
            2,
            100,
            "2.example.com",
            "user",
            1,
            200,
            None,
            None,
            [
                "ssh=1.example.com//id=1.example.com_0//chdir=test//python=",
                "ssh=1.example.com//id=1.example.com_1//chdir=test//python=",
                "ssh=user@2.example.com//id=2.example.com_0//chdir=test//python=",
            ],
        ),
        (
            "1.example.com",
            "",
            2,
            100,
            "2.example.com",
            "user",
            1,
            200,
            200,
            None,
            [
                "ssh=user@2.example.com//id=2.example.com_0//chdir=test//python=",
            ],
        ),
        (
            "1.example.com",
            "",
            2,
            100,
            "2.example.com",
            "user",
            1,
            200,
            None,
            1,
            [
                "ssh=1.example.com//id=1.example.com_0//chdir=test//python=",
                "ssh=user@2.example.com//id=2.example.com_0//chdir=test//python=",
            ],
        ),
        (
            "1.example.com",
            "",
            2,
            100,
            "2.example.com",
            "user",
            1,
            200,
            200,
            1,
            [
                "ssh=user@2.example.com//id=2.example.com_0//chdir=test//python=",
            ],
        ),
        (
            "1.example.com",
            "",
            2,
            100,
            "2.example.com",
            "user",
            1,
            200,
            200,
            1,
            [
                "ssh=user@2.example.com//id=2.example.com_0//chdir=test//python=",
            ],
        ),
    ],
)
# pylint: disable=R0913,W0613
def test_schedule(
    mocked_dsession,
    mocked_group,
    mocked_rsync,
    testdir,
    host1,
    host2,
    user1,
    user2,
    cpu_count1,
    cpu_count2,
    memory1,
    memory2,
    mem_per_process,
    max_processes,
    result,
    request,
):
    """Test scheduling of tests on given nodes."""
    ch1 = mock.Mock()
    ch1.gateway.id = host1
    ch2 = mock.Mock()
    ch2.gateway.id = host2

    node1 = user1 + "@" + host1 if user1 else host1
    node2 = user2 + "@" + host2 if user2 else host2

    mocked_group.return_value.remote_exec.return_value.receive_each.return_value = [
        (
            ch1,
            {
                "cpu_count": cpu_count1,
                "virtual_memory": {"available": memory1 * 1024 ** 2},
            },
        ),
        (
            ch2,
            {
                "cpu_count": cpu_count2,
                "virtual_memory": {"available": memory2 * 1024 ** 2},
            },
        ),
    ]
    params = [
        "--cloud-nodes={0}".format(node1),
        "--cloud-node={0}".format(node2),
        "--cloud-chdir=test",
    ]
    if mem_per_process:
        params.append("--cloud-mem-per-process={0}".format(mem_per_process))
    if max_processes:
        params.append("--cloud-max-processes={0}".format(max_processes))

    testdir.inline_run(*params)
    assert mocked_rsync.call_args[0] == (testdir.tmpdir, "test")
    assert mocked_rsync.return_value.add_target_host.call_args_list == [
        mock.call(
            node1,
        ),
        mock.call(
            node2,
        ),
    ]
    assert mocked_rsync.return_value.send.called
    config = mocked_dsession.call_args[0][0]
    assert all(
        tx.startswith(expected) for tx, expected in zip(config.option.tx, result)
    )
    assert config.option.dist == "load"


pytest_cloud/rsync.py
"""Faster rsync."""

import os
import tempfile
import subprocess

from distutils.spawn import find_executable  # pylint: disable=E0611
import py


def make_reltoroot(roots, args):
    """Non validating make_reltoroot."""
    splitcode = "::"
    res = []
    for arg in args:
        arg = str(arg)
        parts = arg.split(splitcode)
        # pylint: disable=E1101
        fspath = py.path.local(parts[0])
        for root in roots:
            rel_root = fspath.relto(root)
            if rel_root or fspath == root:
                parts[0] = root.basename + "/" + rel_root
                break
        res.append(splitcode.join(parts))
    return res


# pylint: disable=R0902,R0205
class RSync(object):
    """Send a directory structure (recursively) to one or multiple remote filesystems."""

    # pylint: disable=R0913,W0613
    def __init__(
        self,
        sourcedir,
        targetdir,
        verbose=False,
        ignores=None,
        includes=None,
        jobs=None,
        debug=False,
        bwlimit=None,
        ssh_cipher=None,
        **kwargs
    ):
        """Initialize new RSync instance."""
        self.sourcedir = str(sourcedir)
        self.targetdir = str(targetdir)
        self.verbose = verbose
        self.debug = debug
        self.ignores = ignores or []
        self.includes = set(includes or [])
        self.targets = set()
        self.jobs = jobs
        self.bwlimit = bwlimit
        self.ssh_cipher = ssh_cipher

    def get_ignores(self):
        """Get ignores."""
        # pylint: disable=E1101
        return [
            str(py.path.local(ignore).relto(os.path.abspath(".")))
            for ignore in self.ignores
        ]

    def get_includes(self):
        """Get includes."""
        # pylint: disable=E1101
        return [
            str(py.path.local(include).relto(os.path.abspath(".")))
            for include in self.includes
        ]

    def send(self, raises=True):
        """Send a sourcedir to all added targets.

        Flag indicates whether to raise an error or return in case of lack of targets.
        """
        parallel = find_executable("parallel")
        if not parallel:
            raise RuntimeError("parallel is not found.")
        fd_ignores, ignores_path = tempfile.mkstemp()
        fd_includes, includes_path = tempfile.mkstemp()
        fd_ignores = os.fdopen(fd_ignores, "w")
        fd_includes = os.fdopen(fd_includes, "w")
        try:
            fd_ignores.writelines(ignore + "\n" for ignore in self.get_ignores())
            fd_ignores.flush()
            fd_includes.writelines(include + "\n" for include in self.get_includes())
            fd_includes.flush()
            subprocess.call(
                [parallel]
                + (["--verbose"] if self.verbose else [])
                + [
                    "--gnu",
                    "--jobs={0}".format(self.jobs or len(self.targets)),
                    "rsync -arHAXx{verbose} "
                    "{bwlimit}"
                    "--ignore-errors "
                    "--include-from={includes} "
                    "--exclude-from={ignores} "
                    "--numeric-ids "
                    "--force "
                    "--inplace "
                    "--delete-excluded "
                    "--delete "
                    '-e "ssh -T -c {ssh_cipher} -o Compression=no -x" '
                    ". {{}}:{chdir}".format(
                        verbose="v" if self.verbose else "",
                        bwlimit="--bwlimit={0} ".format(self.bwlimit)
                        if self.bwlimit
                        else "",
                        ssh_cipher=self.ssh_cipher,
                        chdir=self.targetdir,
                        ignores=ignores_path,
                        includes=includes_path,
                    ),
                    ":::",
                ]
                + list(self.targets)
            )
        finally:
            fd_ignores.close()
            fd_includes.close()
            os.unlink(ignores_path)
            os.unlink(includes_path)

    def add_target_host(self, host):
        """Add a remote target."""
        self.targets.add(host)


pytest_cloud/patches.py
"""Monkey patches."""

import os
import xdist
from xdist import workermanage

from .rsync import make_reltoroot


# pylint: disable=R0913,W0613
def rsync(self, gateway, source, notify=None, verbose=False, ignores=None):
    """Perform rsync to remote hosts for node."""
    spec = gateway.spec
    if spec.popen and not spec.chdir:
        gateway.remote_exec(
            """
            import sys ; sys.path.insert(0, %r)
        """
            % os.path.dirname(str(source))
        ).waitclose()
        return
    if (spec, source) in self._rsynced_specs:
        return

    self.config.hook.pytest_xdist_rsyncstart(
        source=source,
        gateways=[gateway],
    )
    self.config.hook.pytest_xdist_rsyncfinish(
        source=source,
        gateways=[gateway],
    )


def activate_env(channel, virtualenv_path, develop_eggs=None):
    """Activate virtual environment.

    Executed on the remote side.

    :param channel: execnet channel for communication with master node
    :type channel: execnet.gateway_base.Channel
    :param virtualenv_path: relative path to the virtualenv to activate on the remote test node
    :type virtualenv_path: str
    :param develop_eggs: optional list of python packages to be installed in develop mode
    :type develop_eggs: list
    """
    import os.path  # pylint: disable=W0404,C0415
    import sys  # pylint: disable=W0404,C0415
    import subprocess  # pylint: disable=W0404,C0415
    from itertools import chain  # pylint: disable=W0404,C0415

    subprocess.check_call(["find", ".", "-name", "*.pyc", "-delete"])
    if virtualenv_path:
        if develop_eggs:
            python_script = os.path.abspath(
                os.path.normpath(os.path.join(virtualenv_path, "bin", "python"))
            )
            pip_script = os.path.abspath(
                os.path.normpath(os.path.join(virtualenv_path, "bin", "pip"))
            )
            egg = None
            args = (
                python_script,
                pip_script,
                "install",
                "--no-index",
                "--no-deps",
            ) + tuple(chain.from_iterable([("-e", egg) for egg in develop_eggs]))
            subprocess.check_call(args)

        activate_script = os.path.abspath(
            os.path.normpath(os.path.join(virtualenv_path, "bin", "activate_this.py"))
        )
        exec(open(activate_script).read(), {'__file__': activate_script})


def setup(self):
    """Set up a new test worker."""
    self.log("setting up worker session")
    spec = self.gateway.spec
    args = self.config.args
    if not spec.popen or spec.chdir:
        args = make_reltoroot(self.nodemanager.roots, args)
    option_dict = vars(self.config.option)
    if spec.popen and not spec.via:
        name = "popen-%s" % self.gateway.id
        basetemp = self.config._tmpdirhandler.getbasetemp()
        option_dict["basetemp"] = str(basetemp.join(name))
    self.config.hook.pytest_configure_node(node=self)
    self.channel = self.gateway.remote_exec(xdist.remote)
    if self.putevent:
        self.channel.setcallback(self.process_from_remote, endmarker=self.ENDMARK)
    self.channel.send((self.workerinput, args, option_dict, None))


def apply_patches():
    """Apply monkey patches."""
    workermanage.make_reltoroot = make_reltoroot
    workermanage.NodeManager.rsync = rsync
    workermanage.WorkerController.setup = setup


pytest_cloud/plugin.py
"""Distributed tests planner plugin for pytest testing framework.

Provides an easy way of running tests amoung several test nodes (workers).
"""
from __future__ import division

import argparse
import sys

try:
    from itertools import filterfalse  # pylint: disable=E0611
except ImportError:
    from itertools import ifilterfalse as filterfalse  # pylint: disable=E0611
from itertools import chain
import math
import os.path

import timeout_decorator

import execnet
from xdist.workermanage import (
    NodeManager,
)
import six

import pytest

from .rsync import RSync
from . import patches


# pylint: disable=R0903,R0205
class CloudXdistPlugin(object):
    """Plugin class to defer pytest-xdist hook handler."""


@pytest.mark.trylast
def pytest_configure(config):
    """Register pytest-cloud's deferred plugin."""
    if (
        getattr(config, "workerinput", {}).get("workerid", "local") == "local"
        and config.option.cloud_nodes
        and config.pluginmanager.getplugin("xdist")
    ):
        config.pluginmanager.register(CloudXdistPlugin())


# pylint: disable=W0105
def _ensure_value(namespace, name, value):
    """Ensure value in the namespace. Copied from older version of argparse as is."""
    if getattr(namespace, name, None) is None:
        setattr(namespace, name, value)
    return getattr(namespace, name)


# pylint: disable=R0903
class NodesAction(argparse.Action):
    """Parses out a space-separated list of nodes and extends dest with it."""

    def __call__(self, parser, namespace, values, option_string=None):
        """Parse out space-separated list of nodes."""
        if not hasattr(argparse, "_copy_items"):
            # pylint: disable=E1101
            copy = argparse._copy.copy
        else:
            copy = argparse._copy_items
        items = copy(_ensure_value(namespace, self.dest, []))
        items.extend([value.strip() for value in values.split()])
        setattr(namespace, self.dest, items)


def get_virtualenv_path():
    """Get virtualenv path if test process is using virtual environment inside of current folder."""
    venv_path = os.path.realpath(os.path.dirname(os.path.dirname(sys.executable)))
    if os.path.realpath(os.environ["PWD"]) in venv_path:
        return os.path.relpath(venv_path)
    return None


def pytest_addoption(parser):
    """Pytest hook to add custom command line option(s)."""
    group = parser.getgroup("cloud", "distributed tests scheduler")
    group.addoption(
        "--cloud-python",
        help="python executable name to be used on the remote test nodes."
        "Default is the executable used on the master.",
        type=str,
        action="store",
        dest="cloud_python",
        metavar="NAME",
        default="python{0}.{1}".format(*sys.version_info),
    )
    group._addoption(
        "--cloud-chdir",
        metavar="DIR",
        action="store",
        dest="cloud_chdir",
        default=os.path.join(
            "pytest", os.environ["USER"], os.path.basename(os.environ["PWD"])
        ).replace(os.path.sep, "_"),
        help="relative path on remote node to run tests in. Default is pytest_<username>_<current_folder_name>",
    )
    group.addoption(
        "--cloud-nodes",
        help="space-separated test node list to use for distributed testing",
        type=str,
        action=NodesAction,
        dest="cloud_nodes",
        metavar="'USER@HOST",
        default=[],
    )
    group.addoption(
        "--cloud-node",
        help="test node to use for distributed testing",
        type=str,
        action="append",
        dest="cloud_nodes",
        metavar="USER@HOST",
        default=[],
    )
    group.addoption(
        "--cloud-virtualenv-path",
        help="relative path to the virtualenv to be used on the remote test nodes.",
        type=str,
        action="store",
        dest="cloud_virtualenv_path",
        metavar="PATH",
        default=get_virtualenv_path(),
    )
    group.addoption(
        "--cloud-mem-per-process",
        help="amount of memory roughly needed for test process, in megabytes",
        type=int,
        action="store",
        dest="cloud_mem_per_process",
        metavar="NUMBER",
        default=None,
    )
    group.addoption(
        "--cloud-max-processes",
        help="maximum number of processes per test node",
        type=int,
        action="store",
        dest="cloud_max_processes",
        metavar="NUMBER",
        default=None,
    )
    group.addoption(
        "--cloud-rsync-max-processes",
        help="maximum number of rsync processes",
        type=int,
        action="store",
        dest="cloud_rsync_max_processes",
        metavar="NUMBER",
        default=None,
    )
    group.addoption(
        "--cloud-rsync-bandwidth-limit",
        help="maximum number of processes per test node",
        type=int,
        action="store",
        dest="cloud_rsync_bandwidth_limit",
        metavar="NUMBER",
        default=10000,
    )
    parser.addoption(
        "--cloud-rsync-cipher",
        help="cipher for ssh connection used by rsync",
        type=str,
        dest="cloud_rsync_cipher",
        metavar="STRING",
        default="aes128-gcm@openssh.com",
    )
    parser.addini(
        "cloud_develop_eggs",
        "list of python package paths to install in develop mode on the remote side",
        type="pathlist",
    )


@pytest.mark.tryfirst
def pytest_cmdline_main(config):
    """Apply custom cmd line manipulation for pytest-cloud."""
    check_options(config)


def get_node_capabilities(channel):
    """Get test node capabilities.

    Executed on the remote side.

    :param channel: execnet channel for communication with master node
    :type channel: execnet.gateway_base.Channel

    :return: `dict` in form {'cpu_count': 1, 'virtual_memory': {'available': 100, 'total': 200}}
    :rtype: dict
    """
    import psutil  # pylint: disable=C0415

    memory = psutil.virtual_memory()
    caps = dict(
        cpu_count=psutil.cpu_count(),
        virtual_memory=dict(available=memory.available, total=memory.total),
    )
    channel.send(caps)


# pylint: disable=R0913
def get_node_specs(
    node, host, caps, python=None, chdir=None, mem_per_process=None, max_processes=None
):
    """Get single node specs.

    Executed on the master node side.

    :param node: node name in form <username>@<hostname>
    :type node: str
    :param host: hostname of the node
    :type host: str
    :param python: python executable name to use on the remote side
    :type python: str
    :param chdir: relative path where to run (and sync) tests on the remote side
    :type chdir: str
    :param mem_per_process: optional amount of memory per process needed, in megabytest
    :type mem_per_process: int
    :param max_processes: optional maximum number of processes per test node
    :type max_processes: int

    :return: `list` of test gateway specs for single test node in form ['1*ssh=<node>//id=<hostname>_<index>', ...]
    :rtype: list
    """
    count = min(max_processes or six.MAXSIZE, caps["cpu_count"])
    if mem_per_process:
        count = min(
            int(math.floor(caps["virtual_memory"]["available"] / mem_per_process)),
            count,
        )
    for index in range(count):
        fmt = "ssh={node}//id={host}_{index}//chdir={chdir}//python={python}"
        yield fmt.format(
            count=count, node=node, host=host, index=index, chdir=chdir, python=python
        )


def unique_everseen(iterable, key=None):
    """List unique elements, preserving order. Remember all elements ever seen."""
    # unique_everseen('AAAABBBCCDAABBB') --> A B C D
    # unique_everseen('ABBCcAD', str.lower) --> A B C D
    seen = set()
    seen_add = seen.add
    if key is None:
        for element in filterfalse(seen.__contains__, iterable):
            seen_add(element)
            yield element
    else:
        for element in iterable:
            k = key(element)
            if k not in seen:
                seen_add(k)
                yield element


@timeout_decorator.timeout(3)
def make_gateway(group, spec):
    """Make a gateway."""
    group.makegateway(spec)


def get_develop_eggs(root_dir, config):
    """Get list of eggs to install in develop mode."""
    return [
        "." + os.path.sep + path.relto(root_dir)
        for path in config.getini("cloud_develop_eggs")
    ]


def get_nodes_specs(
    nodes,
    python=None,
    chdir=None,
    virtualenv_path=None,
    mem_per_process=None,
    max_processes=None,
    rsync_max_processes=None,
    rsync_bandwidth_limit=None,
    rsync_cipher=None,
    config=None,
):
    """Get nodes specs.

    Get list of node names, connect to each of them, get the system information, produce the list of node specs out of
    that information filtering non-connectable nodes and nodes which don't comply the requirements.
    Executed on the master node side.

    :param nodes: `list` of node names in form [[<username>@]<hostname>, ...]
    :type nodes: list
    :param python: python executable name to use on the remote side
    :type python: str
    :param chdir: relative path where to run (and sync) tests on the remote side
    :type chdir: str
    :param virtualenv_path: relative path to the virtualenv to activate on the remote test node
    :type virtualenv_path: str
    :param mem_per_process: optional amount of memory per process needed, in megabytest
    :type mem_per_process: int
    :param max_processes: optional maximum number of processes per test node
    :type max_processes: int
    :param rsync_max_processes: optional maximum number of rsync processes
    :type rsync_max_processes: int
    :param rsync_bandwidth_limit: optional bandwidth limit per rsync process in kilobytes per second
    :type rsync_bandwidth_limit: int
    :param config: pytest config object
    :type config: pytest.Config

    :return: `list` of test gateway specs for all test nodes which confirm given requirements
        in form ['1*ssh=<node>//id=<hostname>:<index>', ...]
    :rtype: list
    """
    # pylint: disable=E1101
    group = execnet.Group()
    try:
        n_m = NodeManager(config, specs=[])
        if virtualenv_path:
            virtualenv_path = os.path.relpath(virtualenv_path)
        node_specs = []
        node_caps = {}
        root_dir = config.rootdir
        nodes = list(unique_everseen(nodes))
        print("Detected root dir: {0}".format(root_dir))
        rsync = RSync(
            root_dir,
            chdir,
            includes=config.getini("rsyncdirs"),
            jobs=rsync_max_processes or len(nodes),
            bwlimit=rsync_bandwidth_limit,
            bandwidth_limit=rsync_bandwidth_limit,
            ssh_cipher=rsync_cipher,
            **n_m.rsyncoptions
        )
        print("Detecting connectable test nodes...")
        for node in nodes:
            host = node.split("@")[1] if "@" in node else node
            spec = "ssh={node}//id={host}//chdir={chdir}//python={python}".format(
                node=node, host=host, chdir=chdir, python=python
            )
            try:
                make_gateway(group, spec)
            except Exception:  # pylint: disable=W0703
                continue
            rsync.add_target_host(node)
            node_specs.append((node, host))
        if node_specs:
            print(
                "Found {0} connectable test nodes: {1}".format(
                    len(node_specs), rsync.targets
                )
            )
        else:
            pytest.exit("None of the given test nodes are connectable")
        print("RSyncing directory structure")
        rsync.send()
        print("RSync finished")
        develop_eggs = get_develop_eggs(root_dir, config)
        group.remote_exec(
            patches.activate_env,
            virtualenv_path=virtualenv_path,
            develop_eggs=develop_eggs,
        ).waitclose()
        multi_channel = group.remote_exec(get_node_capabilities)
        try:
            caps = multi_channel.receive_each(True)
            for channel, cap in caps:
                node_caps[channel.gateway.id] = cap
        finally:
            multi_channel.waitclose()
        return list(
            chain.from_iterable(
                get_node_specs(
                    node,
                    hst,
                    node_caps[hst],
                    python=os.path.join(chdir, virtualenv_path, "bin", python),
                    chdir=chdir,
                    mem_per_process=mem_per_process,
                    max_processes=max_processes,
                )
                for node, hst in node_specs
            )
        )
    finally:
        try:
            group.terminate()
        except Exception:  # pylint: disable=W0703
            pass


def check_options(config):
    """Process options to manipulate (produce other options) important for pytest-cloud."""
    if (
        getattr(config, "workerinput", {}).get("workerid", "local") == "local"
        and config.option.cloud_nodes
    ):
        patches.apply_patches()
        mem_per_process = config.option.cloud_mem_per_process
        if mem_per_process:
            mem_per_process = mem_per_process * 1024 * 1024
        virtualenv_path = config.option.cloud_virtualenv_path
        chdir = config.option.cloud_chdir
        python = config.option.cloud_python
        node_specs = get_nodes_specs(
            config.option.cloud_nodes,
            chdir=chdir,
            python=python,
            virtualenv_path=virtualenv_path,
            rsync_max_processes=config.option.cloud_rsync_max_processes,
            rsync_bandwidth_limit=config.option.cloud_rsync_bandwidth_limit,
            max_processes=config.option.cloud_max_processes,
            mem_per_process=mem_per_process,
            rsync_cipher=config.option.cloud_rsync_cipher,
            config=config,
        )
        if node_specs:
            print("Scheduling with {0} parallel test sessions".format(len(node_specs)))
        if not node_specs:
            pytest.exit(
                "None of the given test nodes are able to serve as a test node due to capabilities"
            )
        config.option.tx += node_specs
        config.option.dist = "load"
